src/main/java/com/amazon/opendistroforelasticsearch/sql/rewriter/nestedfield/NestedFieldProjection.java

```diff
@@ -123,23 +217,59 @@ private void buildInnerHit(List<String> fieldNames, NestedQueryBuilder query) {
      * Why linear search? Because NestedQueryBuilder hides "path" field from any access.
      * Assumption: collected NestedQueryBuilder list should be very small or mostly only one.
      */
-    private NestedQueryBuilder findNestedQueryWithSamePath(List<NestedQueryBuilder> nestedQueries, String path) {
+    private NestedQueryBuilder findNestedQueryWithSamePath(
+            List<NestedQueryBuilder> nestedQueries, String path) {
+
         return nestedQueries.stream().
-                             filter(query -> isSamePath(path, query)).
-                             findAny().
-                             orElseGet(createEmptyNestedQuery(path));
+                filter(query -> isSamePath(path, query)).
+                peek(x -> setLastPath(path)).
+                findAny().
+                orElseGet(createNestedQuery(path));
     }
 
     private boolean isSamePath(String path, NestedQueryBuilder query) {
         return nestedQuery(path, query.query(), query.scoreMode()).equals(query);
     }
 
+    /**
+     * groupFieldNamesByPath() uses TreeMap to maintain reverse alphabetical order. Therefore we use previous
+     * path to compare with current path, to help decide if we are done with keeping wrapping nested query
+     * to current nested query.
+     * */
+    private boolean isParentPath(String path, QueryBuilder query) {
+        // query could also be termQueryBuilder, where condition will generate termQueryBuilder
+        if (lastPath.isEmpty() || !(query instanceof NestedQueryBuilder)) {
+            return false;
+        } else {
+            int lastSeparatorIndex = lastPath.lastIndexOf(SEPARATOR);
+            String parentPath = lastSeparatorIndex == -1 ? "" : lastPath.substring(0, lastSeparatorIndex);
+
+            return parentPath.equals(path);
+        }
+    }
+
     /** Create a nested query with match all filter to place inner hits */
-    private Supplier<NestedQueryBuilder> createEmptyNestedQuery(String path) {
+    private Supplier<NestedQueryBuilder> createNestedQuery(String path) {
         return () -> {
-            NestedQueryBuilder nestedQuery = nestedQuery(path, matchAllQuery(), ScoreMode.None);
-            ((BoolQueryBuilder) query().filter().get(0)).must(nestedQuery);
-            return nestedQuery;
+            final NestedQueryBuilder newNestedQuery;
+            List<QueryBuilder> mustList = ((BoolQueryBuilder) query().filter().get(0)).must();
+            int lastIndex = mustList.size() - 1;
+
+            if (mustList.isEmpty() || !isParentPath(path, mustList.get(lastIndex))) {
+                // create empty nested query
+                newNestedQuery = nestedQuery(path, matchAllQuery(), ScoreMode.None);
+            } else {
+                // create a nested query containing another nested query
+                NestedQueryBuilder oldNestedQuery = (NestedQueryBuilder) mustList.get(lastIndex);
```

Comment:
I would love to have more documentation on this method especially when you introduce the `oldNestedQuery` along with `newNestedQuery`. Since we are not all very familiar with all the codepaths, it will be very useful if someone spending time to understand one path will share his/her knowledge in comments.